### ISSUE TITLE
Implement nominal type inheritance

### ISSUE THREAD

**Issue by agoose77** [scikit-hep/awkward#2433]
### Description of new feature

Library authors are currently required to do a reasonable amount of work to extend the types of other libraries. For example, to extend the `vector` types, one must both inherit from the Python classes and register them under new names, but also re-define all of the ufunc overloads in `ak.behavior`.

A related discussion is happening in #2432, whereby a new "typeclass" mechanism is proposed for string-like types. It's my belief #2432 is proposing a solution to the question "is this type like another type". We could implement such logic by defining type inheritance. For example, consider a custom vector type:
```python
ak.behavior["__supertype__", "CustomVector"] = "Vector"
```

Despite the motivations above, I can't think of many concrete use cases, yet. One example would be if the user evaluates `CustomVector + Vector`. Right now, that's an error. With the concept of super and subtypes, the result of `CustomVector + Vector` is now clearly defined; it should invoke the overload for `Vector + Vector`, i.e. return a `Vector`.

This issue is mainly book-keeping; I don't think it's a pressing issue, and I'd want to think it through more before deeming it a sensible direction to take things.

**Comment by jpivarski** [comment]
Here is what one must do to define a "subclass" for a record behavior now:

```python
import numpy as np
import awkward as ak
import vector
vector.register_awkward()


class MomentumRecord4DWithCharge(vector.backends.awkward.MomentumRecord4D):
    pass


class MomentumArray4DWithCharge(vector.backends.awkward.MomentumArray4D):
    pass


ak.behavior["Momentum4DWithCharge"] = MomentumRecord4DWithCharge
ak.behavior["*", "Momentum4DWithCharge"] = MomentumArray4DWithCharge


def add_them(x, y):
    vector_x = ak.with_name(x, "Momentum4D")
    vector_y = ak.with_name(y, "Momentum4D")
    vector_out = ak.with_name(vector_x + vector_y, "Momentum4DWithCharge")
    vector_out["charge"] = x["charge"] + y["charge"]
    return vector_out


ak.behavior[np.add, "Momentum4DWithCharge", "Momentum4DWithCharge"] = add_them
```

Then, given two arrays with the new name:

```python
array1 = ak.zip({
    "px": [[1.1, 2.1], [], [3.1]],
    "py": [[1.2, 2.2], [], [3.2]],
    "pz": [[1.3, 2.3], [], [3.3]],
    "E": [[10.0, 20.0], [], [30.0]],
    "charge": [[+1, -1], [], [-1]],
}, with_name="Momentum4DWithCharge")

array2 = ak.zip({
    "px": [[4.1, 5.1], [], [6.1]],
    "py": [[4.2, 5.2], [], [6.2]],
    "pz": [[4.3, 5.3], [], [6.3]],
    "E": [[40.0, 50.0], [], [60.0]],
    "charge": [[-1, +1], [], [-1]],
}, with_name="Momentum4DWithCharge")

(array1 + array2).show(type=True, limit_cols=120)
```

prints

```
type: 3 * var * Momentum4DWithCharge[
    x: float64,
    y: float64,
    z: float64,
    t: float64,
    charge: int64
]
[[{x: 5.2, y: 5.4, z: 5.6, t: 50, charge: 0}, {x: 7.2, y: 7.4, z: 7.6, t: 70, charge: 0}],
 [],
 [{x: 9.2, y: 9.4, z: 9.6, t: 90, charge: -2}]]
```

Adding new methods and properties is no big deal because this part is already handled by Python class inheritance. `MomentumArray4DWithCharge` is a subclass of `vector.backends.awkward.MomentumArray4D`, so it gets all of the Vector methods and properties:

```python
>>> array1.pt
<Array [[1.63, 3.04], [], [4.46]] type='3 * var * float64'>
```

and I can add any new ones to the class definitions above.

The ugly part is the ufunc-handling, which is most often seen in operator overloading. (Although `__add__`, etc. functions could be added to the classes, there's a standard mapping to NumPy ufuncs, so it is easier to overload the ufuncs and get those operators for free.) The new name `"Momentum4DWithCharge"` does not match any behavior definitions like

```python
ak.behavior[np.add, "Momentum4D", "Momentum4D"] = ...
```

so it's considered a completely new thing, a blank slate. Any attempt to add `array1` and `array2` before defining

```python
ak.behavior[np.add, "Momentum4DWithCharge", "Momentum4DWithCharge"] = add_them
```

would raise

```
TypeError: no numpy.add overloads for custom types: Momentum4DWithCharge, Momentum4DWithCharge

This error occurred while calling

    numpy.add.__call__(
        <MomentumArray4DWithCharge [[{px: 1.1, ...}, ...], ...] type='3 * v...'>
        <MomentumArray4DWithCharge [[{px: 4.1, ...}, ...], ...] type='3 * v...'>
    )
```

which is, admittedly, nicer than just getting the wrong answer (adding them as plain `"Momentum4D"` and losing their charge). This happens because records without explicit ufunc overloads fail in ufuncs: this is what the safety mechanism is supposed to do.

If this were normal Python inheritance, every `__add__`, etc. method would have to be overloaded to say what should happen with the new `"charge"` field. For addition, charges are added; for scalar multiplication, the charge should be unchanged but passed through, for computing the magnitude (`__abs__`), the charge is irrelevant, etc.

In this system, we still have to do that, but all of those overloads are not nested within the class definition, just as the original ufunc implementations were not nested within the class definition. That's not different. But there's no notion of `super()`; we can only say that addition on `"Momentum4DWithCharge"` should do what addition on `"Momentum4D"` does plus a little more by explicitly casting to the supertype. Nothing tells us that Momentum4DWithCharge's direct supertype is Momentum4D.

```python
def add_them(x, y):
    vector_x = ak.with_name(x, "Momentum4D")
    vector_y = ak.with_name(y, "Momentum4D")
    vector_out = ak.with_name(vector_x + vector_y, "Momentum4DWithCharge")
    vector_out["charge"] = x["charge"] + y["charge"]
    return vector_out
```

The `with_name` dance could get annoying after a while.

Another thing that isn't covered is what to do with

```python
ak.behavior[np.add, "Momentum4DWithCharge", "Momentum4D"] = ...
ak.behavior[np.add, "Momentum4D", "Momentum4DWithCharge"] = ...
```

Since they're not specified, these combinations would raise the `TypeError`. They can be explicitly implemented, and the preferred implementation would probably be to upcast the Momentum4DWithCharge into another Momentum4D and proceed with the pair of Momentum4Ds, since the reverse is impossible due to the lack of `"charge"` data. That natural implementation could be implicit, if the exact combination is not found but a superclass is. (We'd have to check all combinations of superclasses, though!)

In summary, inheritance can be implemented now, but it's more manual than it ought to be. Some drudgery is inescapable: you _have to_ say how the new `"charge"` field affects every operation (both ufuncs and methods/properties). If there was a way to look up a named record type's supertype, however, then at least we could provide something like a `super()` function to apply the parent's operation and automatically (dynamically?) fill the matrix of ufunc subclass-superclass argument combinations.

**Comment by jpivarski** [comment]
@Saransh-cpp, this is the original feature request for the behavior-subclassing. I'm pointing this out, not so that you implement the whole thing, but as context for the prototype-like subclassing you're implementing with a utility function.

**Comment by Saransh-cpp** [comment]
Thanks! The comment above was very helpful!
